src/net/connect.cpp

```diff
@@ -157,8 +156,44 @@ void connect_ssl(bufferevent *orig_bev, ssl_st *ssl,
     }
 
     bufferevent_setcb(bev, nullptr, nullptr, mk_bufferevent_on_event,
-            new Callback<bufferevent *>([cb, logger](Error err, bufferevent *bev) {
+            new Callback<bufferevent *>([cb, logger, ssl, hostname](Error err, bufferevent *bev) {
+                int hostname_validate_err = 0;
+
                 logger->debug("connect ssl... callback");
+
+                ssl_st *ssl = bufferevent_openssl_get_ssl(bev);
+                long verify_err = SSL_get_verify_result(ssl);
+                if (verify_err != X509_V_OK) {
+                    debug("ssl: got an invalid certificate");
+                    bufferevent_free(bev);
+                    cb(SSLInvalidCertificateError(X509_verify_cert_error_string(verify_err)), nullptr);
+                    return;
+                }
+
+                if (err) {
+                    logger->debug("error in connection.");
+                    bufferevent_free(bev);
+                    cb(err, nullptr);
+                    return;
+                }
```

Comment:
I'd refactor this `if (err) {` block before the above block, such that first we check for error and then we see what LibreSSL has to tell us. I've checked the code, and in the current code having the blocks are they are is not an error, but in any case I'd be more happy having the check for error at the beginning as I think structuring the code like this is more future proof.